PsfLauncher/main.cpp

```diff
@@ -13,223 +13,227 @@
         }
-        return exitCode;
     }
-    else
+
+    //Launch the end powershell script if we are using one.
+    auto endScriptInformation = PSFQueryEndScriptInfo();
+    if (endScriptInformation)
     {
-        // Non Exe case, use shell launching to pick up local FTA
-        auto nonexePath = packageRoot / exeName;
-
-        CoInitializeEx(NULL, COINIT_APARTMENTTHREADED | COINIT_DISABLE_OLE1DDE);
-        SHELLEXECUTEINFO shex;
-        memset(&shex, 0, sizeof(SHELLEXECUTEINFO));
-        shex.cbSize = sizeof(shex);
-        shex.fMask = SEE_MASK_NOCLOSEPROCESS;
-        shex.hwnd = (HWND)NULL;
-        shex.lpVerb = NULL;  // just use the default action for the fta
-        shex.lpFile = nonexePath.c_str();
-        shex.lpParameters = exeArgString;
-        shex.lpDirectory = dirStr ? (packageRoot / dirStr).c_str() : nullptr;
-        shex.nShow = static_cast<WORD>(cmdShow);
-        
-
-        Log("\tUsing Shell launch: %ls %ls", shex.lpFile, shex.lpParameters);
-        if (!ShellExecuteEx(&shex))
+        ErrorInformation endingScriptError = RunScript(endScriptInformation, packageRoot, dirStr, cmdShow);
+        error.AddExeName(L"Powershell.exe");
+
+        //If there is an existing error from Monitor or the packaged exe
+        if (error.IsThereAnError())
         {
-            std::wostringstream ss;
-            auto err{ ::GetLastError() };
-            // Remove the ".\r\n" that gets added to all messages
-            auto msg = widen(std::system_category().message(err));
-            msg.resize(msg.length() - 3);
-            ss << L"ERROR: Failed to create detoured shell process\n  Path: \"" << exeName << "\"\n  Error: " << msg << " (" << err << ")";
-            ::PSFReportError(ss.str().c_str());
-            return err;
+            ::PSFReportError(error.Print().c_str());
+            return error.GetErrorNumber();
         }
-        else
-        { 
-            // Propagate exit code to caller, in case they care
-            switch (::WaitForSingleObject(shex.hProcess, INFINITE))
-            {
-            case WAIT_OBJECT_0: // SUCCESS
-                if (((unsigned long long)(shex.hInstApp)) > 32)
-                    return 0;
-                else
-                    return (int)(unsigned long long)(shex.hInstApp);
-            case WAIT_FAILED:
-                return ::GetLastError();
-
-            default:
-                return ERROR_INVALID_HANDLE;
-            }			
+        else if (endingScriptError.IsThereAnError())
+        {
+            ::PSFReportError(endingScriptError.Print().c_str());
+            return endingScriptError.GetErrorNumber();
         }
     }
+
+    return 0;
 }
 catch (...)
 {
     ::PSFReportError(widen(message_from_caught_exception()).c_str());
     return win32_from_caught_exception();
+}
 
+ErrorInformation RunScript(const psf::json_object * scriptInformation, std::filesystem::path packageRoot, LPCWSTR dirStr, int cmdShow)

```

Comment:
>* [](start = 50, length = 1)

&

Reply:
I made a comment about this pointer in another comment.  The api in config.cpp returns a pointer to a json_object and I don't want to change the underlying implementation in this PR because I don't know what the far reaching impacts of the change are.

Reply:
again, return values are fine - params are what I'm referring to